docs/src/callbacks.md

```diff
@@ -38,17 +38,26 @@ Callback support is limited to a few solvers. This includes
     underlying solver's callback documentation to understand details specific to
     each solver.
 
-## Information that can be queried during callbacks
+## Things you can and cannot do during callbacks
 
-In a callback, the only thing you may query is the primal value of the
-variables using [`callback_value`](@ref).
+There is a very limited range of things you can do during a callback. Only use 
+the functions and macros explicitly stated in this page of the documentation, or
+in the [Callbacks example](/examples/callbacks).
 
-If you need any other information, use a solver-dependent callback instead.
+Using any other part of the JuMP API (e.g., adding a constraint with [`@constraint`](@ref)
+or modifying a variable bound with [`set_lower_bound`](@ref)) is undefined
+behavior, and your solver may throw an error, return an incorrect solution, or 
+result in a segfault that aborts Julia.
```

Comment:
There shouldn't be any segfault if the user does not use the C API. You don't punish users from not following the rules with segfault, only C/C++ does that and that's why people move to Julia.
Same think for incorrect solution, we should have errors from the wrappers.

Reply:
We can add error messages in some key places, but I don't think that providing this level of guarantees is really feasible. We would have to go through 100% of the JuMP API and add logic to check whether we're currently in a callback and what the current state of the callback is.

Reply:
It should be done in the solvers wrappers, not JuMP. We can warn the user that these are still untested path but we should aim at having unit test for this. All MOI methods already start by checking whether the index is valid and the result index is valid, this would just be another test

Reply:
Wrappers can help by adding extra checks, but I don't think they should be unit tests. Ideally, the underlying solver would object to adding a constraint or variable inside a callback. 

Either way, having this documentation at the JuMP level is a win in the near-term.